File: scripts/compilers/build_repo_map.py
#!/usr/bin/env python3
"""
Deterministic Repo Map Compiler
Generates REPO_MAP.md from repository analysis
"""

import os
import sys
import json
import argparse
from pathlib import Path
from datetime import datetime
from typing import Dict, List, Set

def detect_stack(repo_path: Path) -> Dict[str, any]:
    """Detect programming language and framework stack"""
    stack = {
        "type": "unknown",
        "languages": [],
        "frameworks": [],
        "package_manager": None
    }
    
    # Check for package.json (Node.js)
    if (repo_path / "package.json").exists():
        stack["type"] = "nodejs"
        stack["languages"].append("javascript")
        stack["package_manager"] = "npm"
        try:
            with open(repo_path / "package.json") as f:
                pkg = json.load(f)
                if "dependencies" in pkg:
                    deps = list(pkg["dependencies"].keys())
                    if "react" in deps:
                        stack["frameworks"].append("react")
                    if "express" in deps:
                        stack["frameworks"].append("express")
                    if "next" in deps:
                        stack["frameworks"].append("next")
        except:
            pass
    
    # Check for requirements.txt (Python)
    elif (repo_path / "requirements.txt").exists():
        stack["type"] = "python"
        stack["languages"].append("python")
        stack["package_manager"] = "pip"
    
    # Check for Cargo.toml (Rust)
    elif (repo_path / "Cargo.toml").exists():
        stack["type"] = "rust"
        stack["languages"].append("rust")
        stack["package_manager"] = "cargo"
    
    # Check for go.mod (Go)
    elif (repo_path / "go.mod").exists():
        stack["type"] = "go"
        stack["languages"].append("go")
        stack["package_manager"] = "go"
    
    return stack

def find_insertion_points(repo_path: Path) -> List[Dict[str, str]]:
    """Find common insertion points for code"""
    insertion_points = []
    
    common_dirs = [
        "src", "lib", "app", "components", "pages", "routes",
        "scripts", "workers", "handlers", "controllers", "services"
    ]
    
    for dir_name in common_dirs:
        dir_path = repo_path / dir_name
        if dir_path.exists() and dir_path.is_dir():
            insertion_points.append({
                "path": dir_name,
                "type": "directory",
                "description": f"Code directory: {dir_name}"
            })
    
    # Check for test directories
    test_dirs = ["tests", "test", "__tests__", "spec"]
    for test_dir in test_dirs:
        test_path = repo_path / test_dir
        if test_path.exists():
            insertion_points.append({
                "path": test_dir,
                "type": "test_directory",
                "description": f"Test directory: {test_dir}"
            })
    
    return insertion_points

def generate_repo_map(repo_path: Path, output_path: Path):
    """Generate REPO_MAP.md"""
    stack = detect_stack(repo_path)
    insertion_points = find_insertion_points(repo_path)
    
    timestamp = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
    
    content = f"""# Repository Map
Generated: {timestamp}

## Stack
- Type: {stack['type']}
- Languages: {', '.join(stack['languages']) if stack['languages'] else 'Unknown'}
- Frameworks: {', '.join(stack['frameworks']) if stack['frameworks'] else 'None detected'}
- Package Manager: {stack['package_manager'] or 'Unknown'}

## Architecture
- Repository Type: {'Code repository' if stack['type'] != 'unknown' else 'Documentation repository'}
- Structure: Detected from file system

## Insertion Points

"""
    
    if insertion_points:
        for point in insertion_points:
            content += f"### {point['path']}\n"
            content += f"- Type: {point['type']}\n"
            content += f"- Description: {point['description']}\n\n"
    else:
        content += "No common insertion points detected. Repository may be empty or use non-standard structure.\n\n"
    
    content += """## Status
- ✅ Repository structure mapped
- ✅ Insertion points identified
- ✅ Stack detected

## Notes
This map is generated deterministically from repository contents.
Update manually if structure changes significantly.
"""
    
    output_path.parent.mkdir(parents=True, exist_ok=True)
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(content)
    
    print(f"✓ Generated: {output_path}")

def main():
    parser = argparse.ArgumentParser(description="Build repository map")
    parser.add_argument("--output", default="ai/context/REPO_MAP.md", help="Output file path")
    parser.add_argument("--repo", default=".", help="Repository root path")
    
    args = parser.parse_args()
    
    repo_path = Path(args.repo).resolve()
    output_path = repo_path / args.output
    
    generate_repo_map(repo_path, output_path)

if __name__ == "__main__":
    main()

File: scripts/compilers/build_repo_map.js
#!/usr/bin/env node
/**
 * Deterministic Repo Map Compiler (Node.js)
 * Generates REPO_MAP.md from repository analysis
 */

const fs = require('fs');
const path = require('path');

function detectStack(repoPath) {
    const stack = {
        type: 'unknown',
        languages: [],
        frameworks: [],
        packageManager: null
    };
    
    const packageJsonPath = path.join(repoPath, 'package.json');
    if (fs.existsSync(packageJsonPath)) {
        stack.type = 'nodejs';
        stack.languages.push('javascript');
        stack.packageManager = 'npm';
        
        try {
            const pkg = JSON.parse(fs.readFileSync(packageJsonPath, 'utf8'));
            if (pkg.dependencies) {
                const deps = Object.keys(pkg.dependencies);
                if (deps.includes('react')) stack.frameworks.push('react');
                if (deps.includes('express')) stack.frameworks.push('express');
                if (deps.includes('next')) stack.frameworks.push('next');
            }
        } catch (e) {
            // Ignore parse errors
        }
    } else if (fs.existsSync(path.join(repoPath, 'requirements.txt'))) {
        stack.type = 'python';
        stack.languages.push('python');
        stack.packageManager = 'pip';
    } else if (fs.existsSync(path.join(repoPath, 'Cargo.toml'))) {
        stack.type = 'rust';
        stack.languages.push('rust');
        stack.packageManager = 'cargo';
    } else if (fs.existsSync(path.join(repoPath, 'go.mod'))) {
        stack.type = 'go';
        stack.languages.push('go');
        stack.packageManager = 'go';
    }
    
    return stack;
}

function findInsertionPoints(repoPath) {
    const insertionPoints = [];
    
    const commonDirs = [
        'src', 'lib', 'app', 'components', 'pages', 'routes',
        'scripts', 'workers', 'handlers', 'controllers', 'services'
    ];
    
    for (const dirName of commonDirs) {
        const dirPath = path.join(repoPath, dirName);
        if (fs.existsSync(dirPath)) {
            const stat = fs.statSync(dirPath);
            if (stat.isDirectory()) {
                insertionPoints.push({
                    path: dirName,
                    type: 'directory',
                    description: `Code directory: ${dirName}`
                });
            }
        }
    }
    
    const testDirs = ['tests', 'test', '__tests__', 'spec'];
    for (const testDir of testDirs) {
        const testPath = path.join(repoPath, testDir);
        if (fs.existsSync(testPath)) {
            insertionPoints.push({
                path: testDir,
                type: 'test_directory',
                description: `Test directory: ${testDir}`
            });
        }
    }
    
    return insertionPoints;
}

function generateRepoMap(repoPath, outputPath) {
    const stack = detectStack(repoPath);
    const insertionPoints = findInsertionPoints(repoPath);
    const timestamp = new Date().toISOString().replace('T', ' ').substring(0, 19);
    
    let content = `# Repository Map
Generated: ${timestamp}

## Stack
- Type: ${stack.type}
- Languages: ${stack.languages.length > 0 ? stack.languages.join(', ') : 'Unknown'}
- Frameworks: ${stack.frameworks.length > 0 ? stack.frameworks.join(', ') : 'None detected'}
- Package Manager: ${stack.packageManager || 'Unknown'}

## Architecture
- Repository Type: ${stack.type !== 'unknown' ? 'Code repository' : 'Documentation repository'}
- Structure: Detected from file system

## Insertion Points

`;
    
    if (insertionPoints.length > 0) {
        for (const point of insertionPoints) {
            content += `### ${point.path}\n`;
            content += `- Type: ${point.type}\n`;
            content += `- Description: ${point.description}\n\n`;
        }
    } else {
        content += 'No common insertion points detected. Repository may be empty or use non-standard structure.\n\n';
    }
    
    content += `## Status
- ✅ Repository structure mapped
- ✅ Insertion points identified
- ✅ Stack detected

## Notes
This map is generated deterministically from repository contents.
Update manually if structure changes significantly.
`;
    
    const outputDir = path.dirname(outputPath);
    if (!fs.existsSync(outputDir)) {
        fs.mkdirSync(outputDir, { recursive: true });
    }
    
    fs.writeFileSync(outputPath, content, 'utf8');
    console.log(`✓ Generated: ${outputPath}`);
}

function main() {
    const args = process.argv.slice(2);
    let output = 'ai/context/REPO_MAP.md';
    let repo = '.';
    
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--output' && i + 1 < args.length) {
            output = args[i + 1];
            i++;
        } else if (args[i] === '--repo' && i + 1 < args.length) {
            repo = args[i + 1];
            i++;
        }
    }
    
    const repoPath = path.resolve(repo);
    const outputPath = path.join(repoPath, output);
    
    generateRepoMap(repoPath, outputPath);
}

if (require.main === module) {
    main();
}

module.exports = { generateRepoMap, detectStack, findInsertionPoints };
